### DISCUSSION TITLE
Server should provide default method to throw error

### DISCUSSION THREAD

**Discussion by marswong** [fastify/fastify#2869] [Ideas]
Although there is [fastify-error](https://github.com/fastify/fastify-error), its way to throw an error isn't elegant and simple enough. By contrast, [koa](https://koajs.com/#ctx-throw-status-msg-properties-) handles error in a better way. Imagine a scenario where we have to throw a common HTTP error (e.g: 501 Not Implemented):

- **the fastify-error style**
```js
const createError = require('fastify-error')
const CustomError = createError('ERROR_CODE', 'Not Implemented', 501)
throw new CustomError()
```

- **the koa style**
```js
fastify.throw(501)
```

Which style would most people prefer? Obviously the latter.

**Comment by jsumners** [comment]
`fastify-error` is intended for more robust errors. If you want to throw a simple error with an HTTP status code:

```js
const httpErrors = require('http-errors')

module.exports = function myHandler (req, res) {
  res.send(new httpErrors.NotImplemented())
```

**Reply by marswong** [reply]
good point. There's another simpler way to implement this without extra external modules, 'caz the `http` module already provide http status with message:

```js
const http = require('http')

function BaseError(status, code, message) {
  Error.captureStackTrace(this, BaseError)
  this.name = 'BaseError'
  this.statusCode = status
  this.code = code
  this.message = message
}

async function ErrorPlugin(fastify) {
  fastify.decorate('throw', (status = 500, code = 'FST_ERR_UNKNOWN', message = '') => {
    throw new BaseError(status, code, message || http.STATUS_CODES[status])
  })
}
```

thus, we just need to call `fastify.throw(501)`

**Comment by Eomm** [comment]
`fastify-sensible` lets you write

```
return reply.notFound('Missing Foo Bar')
```

Anyway the `fastify.throw()` feature could be archived with a plugin as well 👍

**Reply by marswong** [reply]
yup, I've already written a plugin in my project, but it'd be better to be built in fastify, doesn't it?

**Reply by jsumners** [reply]
> yup, I've already written a plugin in my project, but it'd be better to be built in fastify, doesn't it?

Only if significant value can be shown. If something can be a plugin, it should be a plugin.

**Reply by Eomm** [reply]
Yes, usually we would like to keep the core as light as possible and let the community build plugins to improve productivity

For example, the main issue to design a `.throw()` interface would be:

- should the `error handler` be triggered or is it a managed error? [see here](https://github.com/fastify/fastify/blob/master/docs/Lifecycle.md#reply-lifecycle)
- in the handler sync function, the `throw new Error()` is an anti-pattern, how could we deal with it?

Right now this is a choice of the user, that will choose a plugin that fits the application's need
